cache/image_cache.go

```diff
@@ -4,29 +4,24 @@ import (
 	"encoding/json"
 	"io"
 
+	"github.com/buildpack/imgutil"
 	"github.com/pkg/errors"
 
-	"github.com/buildpack/lifecycle/image"
 	"github.com/buildpack/lifecycle/metadata"
 )
 
-//go:generate mockgen -package testmock -destination testmock/image_factory.go github.com/buildpack/lifecycle/cache ImageFactory
-type ImageFactory interface {
-	NewEmptyLocal(string) image.Image
-}
-
 type ImageCache struct {
-	factory   ImageFactory
-	origImage image.Image
-	newImage  image.Image
+	initializer func(string) imgutil.Image
+	origImage   imgutil.Image
+	newImage    imgutil.Image
 }
 
-func NewImageCache(factory ImageFactory, origImage image.Image) *ImageCache {
-	newImage := factory.NewEmptyLocal(origImage.Name())
+func NewImageCache(origImage imgutil.Image, initializer func(string) imgutil.Image) *ImageCache {
```

Comment:
I think we should just inject the new empty image instead of injecting this function. It seems like it exists to enable generating a new image after Save happens, but the behavior there feel unintuitive to me. If I add x, and y layers to my cache object, then save, then add z layer to the same cache object I would expect the cache to have x, y, and z layers. Right now I think we would only end up with z layer in that final cache.

Reply:
@ekcasey I think that our cache doesn't work that way be design. Our cache is totally destructive unless you mark things you'd like to keep. In your example, I'd have to "reuse" x and y after my initial commit if wished to continue seeing x any y. I feel like the cache fell into this design due to the underlying design of the image library. To get different behavior, I think we might have to make a larger set of changes. 

I'm wondering if there's a solution where, for now, caches in lifecycle can only be used once per invocation of cacher or restorer.